geode-core/src/main/java/org/apache/geode/distributed/internal/membership/gms/MembershipLocatorBuilderImpl.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.geode.distributed.internal.tcpserver.TcpSocketCreator;
+import org.apache.geode.internal.serialization.ObjectDeserializer;
+import org.apache.geode.internal.serialization.ObjectSerializer;
+
+public final class MembershipLocatorBuilderImpl<ID extends MemberIdentifier> implements
+    MembershipLocatorBuilder<ID> {
+  private int port = 0;
+  private InetAddress bindAddress = null;
+  private ProtocolChecker protocolChecker = (socket, input, firstByte) -> false;
+  private TcpHandler fallbackHandler = new TcpHandlerNoOp();
+  private MembershipLocatorStatistics locatorStats = new MembershipLocatorStatisticsNoOp();
+  private boolean locatorsAreCoordinators = true;
+  private TcpSocketCreator socketCreator;
+  private ObjectSerializer objectSerializer;
+  private ObjectDeserializer objectDeserializer;
+  private Path workingDirectory;
+  private MembershipConfig config;
+  private Supplier<ExecutorService> executorServiceSupplier;
+
+  @Override
+  public MembershipLocatorBuilder<ID> setPort(int port) {
+    this.port = port;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setBindAddress(InetAddress bindAddress) {
+    this.bindAddress = bindAddress;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setProtocolChecker(ProtocolChecker protocolChecker) {
+    this.protocolChecker = protocolChecker;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setExecutorServiceSupplier(
+      Supplier<ExecutorService> executorServiceSupplier) {
+    this.executorServiceSupplier = executorServiceSupplier;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setSocketCreator(TcpSocketCreator socketCreator) {
+    this.socketCreator = socketCreator;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setObjectSerializer(ObjectSerializer objectSerializer) {
+    this.objectSerializer = objectSerializer;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setObjectDeserializer(ObjectDeserializer objectDeserializer) {
+    this.objectDeserializer = objectDeserializer;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setFallbackHandler(TcpHandler fallbackHandler) {
+    this.fallbackHandler = fallbackHandler;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setLocatorsAreCoordinators(boolean locatorsAreCoordinators) {
+    this.locatorsAreCoordinators = locatorsAreCoordinators;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setLocatorStats(MembershipLocatorStatistics locatorStats) {
+    this.locatorStats = locatorStats;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setWorkingDirectory(Path workingDirectory) {
+    this.workingDirectory = workingDirectory;
+    return this;
+  }
+
+  @Override
+  public MembershipLocatorBuilder<ID> setConfig(MembershipConfig config) {
```

Comment:
What attributes are optional with this builder?   It looks like MembershipLocatorImpl will throw an NPE if a MembershipConfig isn't supplied.

Reply:
Anything that doesn't have a default is required. Are you suggesting some additional documentation or other way of indicating what attributes are optional?

Reply:
I guess it would be less error prone if the builder constructor took all of the required settings and then you could set the optional ones with setXYZ() methods.